File: examples/camera_eo_set_shutter_speed.cpp
#include "stdio.h"
#include <pthread.h>
#include <cstdlib>

#include "payloadSdkInterface.h"

#if (CONTROL_METHOD == CONTROL_UART)
T_ConnInfo s_conn = {
    CONTROL_UART,
    payload_uart_port,
    payload_uart_baud
};
#else
T_ConnInfo s_conn = {
    CONTROL_UDP,
    udp_ip_target,
    udp_port_target
};
#endif

PayloadSdkInterface* my_payload = nullptr;

void onPayloadParamChanged(int event, char* param_char, double* param_double);
void quit_handler(int sig);

int main(int argc, char *argv[]){
    printf("Starting SetShutterSpeed example...\n");
    signal(SIGINT, quit_handler);

    // create payloadsdk object
    my_payload = new PayloadSdkInterface(s_conn);

    // init payload
    my_payload->sdkInitConnection();
    printf("Waiting for payload signal! \n");

    // register callback function
    my_payload->regPayloadParamChanged(onPayloadParamChanged);

    // check connection
    my_payload->checkPayloadConnection();

    // Only handle for VIO/ZIO where shutter param is defined
    #if defined VIO || defined ZIO
        printf("[EO] Forcing view source to EO...\n");
	    my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, PAYLOAD_CAMERA_VIEW_EO, PARAM_TYPE_UINT32);
        
        // Read current shutter speed only (by ID)
        my_payload->getPayloadCameraSettingByID((char*)PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED);
        usleep(1000000);

        // Ensure exposure mode allows shutter control
        my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_AUTO_EXPOSURE, PAYLOAD_CAMERA_VIDEO_EXPOSURE_SHUTTER, PARAM_TYPE_UINT32);
        usleep(1000000);
    #endif /* VIO || ZIO */

    while(true)
    {
        // Only handle for VIO/ZIO where shutter param is defined
        #if defined VIO || defined ZIO
            // Set shutter speed to 1/1000 (change constant as needed)
            my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED, PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_10, PARAM_TYPE_UINT32);
            usleep(1000000);

            // Read back shutter speed only (by ID)
            my_payload->getPayloadCameraSettingByID((char*)PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED);
            usleep(1000000);

            // Set shutter speed to 1/1000 (change constant as needed)
            my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED, PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1000, PARAM_TYPE_UINT32);
            usleep(1000000);

            // Read back shutter speed only (by ID)
            my_payload->getPayloadCameraSettingByID((char*)PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED);
            usleep(1000000);
        #else
            exit(-1);
        #endif /* VIO || ZIO */
    }

    // close payload interface
    try {
        my_payload->sdkQuit();
    }
    catch (int error) {}

    return 0;
}

void quit_handler(int sig){
    printf("\n");
    printf("TERMINATING AT USER REQUEST \n");
    printf("\n");

    // close payload interface
    try {
        my_payload->sdkQuit();
    }
    catch (int error) {}

    // end program here
    exit(0);
}

static const char* shutter_value_to_label(int value){
    switch(value){
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_10: return "1/10";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_20: return "1/20";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_50: return "1/50";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_100: return "1/100";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_125: return "1/125";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_500: return "1/500";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_725: return "1/725";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1000: return "1/1000";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1500: return "1/1500";
        case PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_2000: return "1/2000";
        default: return "unknown";
    }
}

void onPayloadParamChanged(int event, char* param_char, double* param){
    switch(event){
    case PAYLOAD_CAM_PARAMS:{
        // Only log shutter speed updates
        if(strcmp(param_char, PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED) == 0){
            int code = static_cast<int>(param[1]);
            printf("Shutter (C_V_SP): %s (code %d)\n", shutter_value_to_label(code), code);
        }
        break;
    }
    default: break;
    }
}

File: examples/camera_eo_set_shutter_speed.py
#!/usr/bin/env python3
import sys
import os

# Add the libs directory to the path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'libs'))

# Import config first to setup environment automatically
from config import config

import time
import signal
import sys
from payload_sdk import PayloadSdkInterface, payload_status_event_t, payload_param_t, PAYLOAD_TYPE
from payload_define import *
from pymavlink import mavutil

my_payload = None
time_to_exit = False

# Signal handler for quitting
def quit_handler(sig, frame):
    global my_payload, time_to_exit
    print("\nTERMINATING AT USER REQUEST")
    time_to_exit = True

    # Close payload interface
    if my_payload:
        try:
            my_payload.sdkQuit()
        except Exception as e:
            print(f"Error while quitting payload: {e}")
    # End program 
    sys.exit(0)

# Shutter value to label conversion
def shutter_value_to_label(value):
    shutter_labels = {
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_10: "1/10",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_20: "1/20",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_50: "1/50",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_100: "1/100",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_125: "1/125",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_500: "1/500",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_725: "1/725",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1000: "1/1000",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1500: "1/1500",
        payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_2000: "1/2000"
    }
    return shutter_labels.get(value, "unknown")

# Callback function for payload status changes
def onPayloadParamChanged(event: int, param_char:str ,param: list):
    if payload_status_event_t(event) == payload_status_event_t.PAYLOAD_CAM_PARAMS:
        # Only log shutter speed updates
        if param_char == PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED:
            code = int(param[1])
            print(f"Shutter (C_V_SP): {shutter_value_to_label(code)} (code {code})")

def main():
    global my_payload, time_to_exit

    print("Starting Set gimbal mode example...")
    signal.signal(signal.SIGINT, quit_handler)

    # Create payloadsdk object
    my_payload = PayloadSdkInterface()

    # Init payload
    my_payload.sdkInitConnection()
    print("Waiting for payload signal!")

    # Register callback function
    my_payload.regPayloadParamChanged(onPayloadParamChanged)

    # Check connection
    my_payload.checkPayloadConnection()

    # Set view source to EO
    print("Set view source to EO!")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, payload_camera_view_src.PAYLOAD_CAMERA_VIEW_EO, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(1) 

    # Only handle for VIO/ZIO where shutter param is defined
    if PAYLOAD_TYPE in ["VIO", "ZIO"]:
 
        # Read current shutter speed only (by ID)
        my_payload.getPayloadCameraSettingByID(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED)
        time.sleep(1)

        # Ensure exposure mode allows shutter control
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_AUTO_EXPOSURE, payload_camera_video_auto_exposure.PAYLOAD_CAMERA_VIDEO_EXPOSURE_SHUTTER, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(1)

        while not time_to_exit:
            # Set shutter speed to 1/1000 (change constant as needed)
            my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED, payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_10, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
            time.sleep(1)

            # Read back shutter speed only (by ID)
            my_payload.getPayloadCameraSettingByID(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED)
            time.sleep(1)

            # Set shutter speed to 1/1000 (change constant as needed)
            my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED, payload_camera_video_shutter_speed.PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED_1_1000, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
            time.sleep(1)

            # Read back shutter speed only (by ID)
            my_payload.getPayloadCameraSettingByID(PAYLOAD_CAMERA_VIDEO_SHUTTER_SPEED)
            time.sleep(1)

    else:
        sys.exit(-1)        

    # Close payload interface
    try:
        my_payload.sdkQuit()
    except Exception as e:
        print(f"Error while quitting payload: {e}")

if __name__ == "__main__":
    main()